rfcs/plugins.md

```diff
@@ -0,0 +1,349 @@
+    `processConfig` as its first argument and a `done` function as its
+    second argument. It should execute its task, calling `done` with an error if something goes wrong.
+    - Note: it should *not* call `process.exit`.
+    - Mandatory option, since this is the entire point of the command.
+
+Note that the command will not ever have `activate` or `deactivate`
+called, those functions are for server plugins.
+
+### Config
+
+The `config` plugin property specifies which config options will be
+forwarded from the config file section for the plugin. The value
+should be a
+[Joi validator](https://github.com/hapijs/joi/blob/master/API.md) to
+validates the config options.
+
+So a `config` key like:
+
+```js
+config: {
+   someKey: Joi.boolean(),
+   some_other_key: Joi.string(),
+   another: Joi.string().default('foo'),
+}
+```
+
+Will allow the following section in `config.toml`:
+
+```toml
+[plugin.my-plugin-name]
+someKey = true
+some_other_key = 'some kinda string'
+```
+
+The values from this config section will be passed to both the
+`processConfig` function for commands, as well as the
+[`activate`](#activate) function for server plugins. The object passed
+will look like:
+
+```js
+{ someKey: true,
+  some_other_key: 'some_kinda_string',
+  another: 'foo',
+}
+```
+
+### HTTP route
+
+The `httpRoute` plugin property is an object with a single key
+defining the http route the plugin will receive requests from. The
+value is the request handler for the route. The function receives
+three arguments:
+ - An Express `request` object
+ - An Express `response` object
+ - A state object of whatever kind is returned by the plugin's
+   [activate](#activate) function
+
+The handler should deal with whatever HTTP methods come to the route.
+
+Example:
+
+```js
+httpRoute: {
+    'hello-world': (req, res, state) => {
+        res.send(`Hello ${state.value}`);
+            res.end()
+    },
+}
+```
+
+Notes:
+ - The handler is responsible for calling `res.end()`.
+ - The route is a hard-prefix from the root of the horizon server address.
+  - It cannot clash with any other other loaded plugin routes (no
+    overlapping)
+      - If multiple routes need to be handled, the r
+      - Example, if you want to create a REST plugin, you need to
+        specify the route as something like `rest` and the endpoints
+        underneath it like `/rest/users` `/rest/vehicles` etc.
+  - It cannot clash with the route horizon uses itself (usually
+    `"horizon"`)
+ - In the configuration, the user may override the httpRoute for the
+   plugin, and the plugin must behave properly.
+ - Express 4.0 api for request and response objects
+
+### Processes
+
+The `processes` plugin property is an object with keys that correspond
+to process titles, and values which are objects with 2 properties:
+ - `run`: a function receiving the plugin state that runs as the main
+   thread of the process. It should not spawn processes itself
```

Comment:
I'm a little confused by this interface.  If I want to e.g. start a RethinkDB proxy in the background, how would I do that except by spawning a process in `run`?